README.md
# S3-Backed-FTP Server

An ftp/sftp server using s3fs to mount an external s3 bucket as ftp/sftp storage.

Started from [existing solution](http://cloudacademy.com/blog/s3-ftp-server/).

## Usage

To run:

1. Copy docker-compose.example.yml to docker-compose.yml and replace the variables used with your values
    - Add users to `USERS` environment variable. These should be listed in the form `username:hashedpassword`, each separated by a space.
     - Passwords for those users should be hashed. There are several ways to hash a user password. A common way is to execute a command like the following: `openssl passwd -crypt {your_password}`. Substitute `{your_password}` with the one you want to hash.
     - You may also use non-hashed passwords if storing passwords in plaintext is fine. To do this, change line ` echo $u | chpasswd -e ` => ` echo $u | chpasswd ` in the `users.sh` file (line #24).
    - Specify the S3 buckets were the files (`FTP_BUCKET`) and configs (`CONFIG_BUCKET`) will be stored.
    - If you are running this container inside an AWS EC2 Instance with an assigned IAM_ROLE, then specify its name in the `IAM_ROLE` environment variable.
    - If you do not have an IAM_ROLE attached to your EC2 Instance or wherever you are running this, then you have to specify the AWS credentials that will be used to access S3. These are the `AWS_ACCESS_KEY_ID` and the `AWS_SECRET_KEY_ID` keys.

2. If you have changed other files aside the `docker-compose.yml` file, then you have to build the docker container using:

    - `docker build --rm -t <docker/tag> path/to/dockerfile/folder`

3. Then after building the container (if necessary), run using:

    docker-compose up -d

   
## Environment Variables

| Variable | Required | Default value | Content
|----------|----------|---------------|----------
| `USERS`         | Yes |             | List of users to add to the ftp/sftp server. Listed in the form username:hashedpassword, each separated by a space.
| `FTP_BUCKET`    | Yes |             | S3 bucket where ftp/sftp users data will be stored.
| `CONFIG_BUCKET` | Yes |             | S3 bucket where the config data (env.list file) will be stored.
| `IAM_ROLE`      | No  |             |  name of role account linked to EC2 instance the container is running in.
| `TRACE`	  | No  | 0           | 1 to enable verbose options in bash scripts
| `MP_UMASK`      | No  | 0022        | option in s3fs
| `UMASK`         | No  | 0002        | option in s3fs, default is rwxrwxr-x

### Optional Environment Variables
These two environment variables only need to be set if there is no linked IAM role to the EC2 instance.

1. ` AWS_ACCESS_KEY_ID ` = IAM user account access key.
2. ` AWS_SECRET_ACCESS_KEY ` = IAM user account secret access key.
3. ` AWS_DEFAULT_REGION ` = bucket region


### GitLab & AWS Elastic Container Registry configuration

To use gitlab CI add the following variables to GitLab CI Configuration:

1. ` AWS_ACCESS_KEY_ID ` = IAM user account access key.
2. ` AWS_SECRET_ACCESS_KEY ` = IAM user account secret access key.
3. ` AWS_ECR_IMAGE ` = Image name in ECR


**Enjoy!**


CHANGELOG.md
# Change Log

## [1.2](https://github.com/opencontentcoop/s3-backed-ftp/tree/1.2) (2019-04-09)

[Minor fix to reduce logging](https://github.com/opencontentcoop/s3-backed-ftp/compare/1.1...1.2)

## [1.1](https://github.com/opencontentcoop/s3-backed-ftp/tree/1.1) (2019-04-09)

[Fixed unpredictable erros on s3fs mount](https://github.com/opencontentcoop/s3-backed-ftp/compare/1.0...1.1)

## [1.0](https://github.com/opencontentcoop/s3-backed-ftp/tree/1.0) (2019-03-04)


\* *This Change Log was automatically generated by [github_changelog_generator](https://github.com/skywinder/Github-Changelog-Generator)*


users.sh
#!/bin/bash
# vim: autoindent tabstop=2 shiftwidth=2 expandtab softtabstop=2 filetype=bash fileencoding=utf-8

[[ $TRACE ]] && set -x

FTP_DIRECTORY="/home/aws/s3bucket/ftp-users"

# Create a group for ftp users
groupadd ftpaccess


# Create a directory where all ftp/sftp users home directories will go
mkdir -p $FTP_DIRECTORY
chown root:root $FTP_DIRECTORY
chmod 755 $FTP_DIRECTORY

# Expecing an environment variable called USERS to look like "bob:hashedbobspassword steve:hashedstevespassword"
for u in $USERS; do
  
  read username passwd <<< $(echo $u | sed 's/:/ /g')

  # User needs to be created every time since stopping the docker container gets rid of users.
  useradd -d "$FTP_DIRECTORY/$username" -s /usr/sbin/nologin $username
  usermod -G ftpaccess $username

  # set the users password
  echo $u | chpasswd -e
  
  if [ -z "$username" ] || [ -z "$passwd" ]; then
    echo "Invalid username:password combination '$u': please fix to create '$username'"
    continue
  elif [ -d "$FTP_DIRECTORY/$username" ]; then
    echo "Skipping creation of '$username' directory: already exists"

    # Directory exists but permissions for it have to be setup anyway.
    chown root:ftpaccess "$FTP_DIRECTORY/$username"
    chmod 750 "$FTP_DIRECTORY/$username"
    chown $username:ftpaccess "$FTP_DIRECTORY/$username/files"
    chmod 750 "$FTP_DIRECTORY/$username/files"
  else
    echo "Creating '$username' directory..."
    
    # Root must own all directories leading up to and including users home directory
    mkdir -p "$FTP_DIRECTORY/$username"
    chown root:ftpaccess "$FTP_DIRECTORY/$username"
    chmod 750 "$FTP_DIRECTORY/$username"
    
    # Need files sub-directory for SFTP chroot
    mkdir -p "$FTP_DIRECTORY/$username/files"
    chown $username:ftpaccess "$FTP_DIRECTORY/$username/files"
    chmod 750 "$FTP_DIRECTORY/$username/files"
  fi
  
done


add_users_in_container.sh
#!/bin/bash
# vim: autoindent tabstop=2 shiftwidth=2 expandtab softtabstop=2 filetype=bash fileencoding=utf-8

[[ $TRACE ]] && set -x

# This script will update the env.list file (file containing USERS environrment variable) and add the new users if there are any.
# Will check for new users at a given time interval (change sleep duration on line 33)

FTP_DIRECTORY="/home/aws/s3bucket/ftp-users"
CONFIG_FILE="env.list" # May need to modify config file name to reflect future changes in env file location/name
SLEEP_DURATION=60
# Change theses next two variables to set different permissions for files/directories
# These were default from vsftpd so change accordingly if necessary
FILE_PERMISSIONS=644
DIRECTORY_PERMISSIONS=755

add_users() {
  aws s3 cp s3://$CONFIG_BUCKET/$CONFIG_FILE ~/$CONFIG_FILE
  USERS=$(cat ~/"$CONFIG_FILE" | grep USERS | cut -d '=' -f2)

  for u in $USERS; do
    read username passwd <<< $(echo $u | sed 's/:/ /g')

    # If account exists set password again 
    # In cases where password changes in env file
    if getent passwd "$username" >/dev/null 2>&1; then
      echo $u | chpasswd -e

      # Fix for issue when pulling files that were uploaded directly to S3 (through aws web console)
      # Permissions when uploaded directly through S3 Web client were set as:
      # 000 root:root
      # This would not allow ftp users to read the files
      
      # Search for files and directories not owned correctly
      find "$FTP_DIRECTORY"/"$username"/files/* \( \! -user "$username" \! -group "$username" \) -print0 | xargs -0 chown "$username:$username"

      # Search for files with incorrect permissions
      find "$FTP_DIRECTORY"/"$username"/files/* -type f \! -perm "$FILE_PERMISSIONS" -print0 | xargs -0 chmod "$FILE_PERMISSIONS"

      # Search for directories with incorrect permissions
      find "$FTP_DIRECTORY"/"$username"/files/* -type d \! -perm "$DIRECTORY_PERMISSIONS" -print0 | xargs -0 chmod "$DIRECTORY_PERMISSIONS"

    fi

    # If user account doesn't exist create it 
    # As well as their home directory 
    if ! getent passwd "$username" >/dev/null 2>&1; then
       useradd -d "$FTP_DIRECTORY/$username" -s /usr/sbin/nologin $username
       usermod -G ftpaccess $username

       mkdir -p "$FTP_DIRECTORY/$username"
       chown root:ftpaccess "$FTP_DIRECTORY/$username"
       chmod 750 "$FTP_DIRECTORY/$username"

       mkdir -p "$FTP_DIRECTORY/$username/files"
       chown $username:ftpaccess "$FTP_DIRECTORY/$username/files"
       chmod 750 "$FTP_DIRECTORY/$username/files"
     fi
   done
}

while true; do
  if [[ -d $FTP_DIRECTORY ]]; then
    add_users
  else
    echo "Missing '$FTP_DIRECTORY', cannot add users, just sleep for $SLEEP_DURATION seconds and retry later"
  fi
  sleep $SLEEP_DURATION
done


s3-fuse.sh
#!/bin/bash
# vim: autoindent tabstop=2 shiftwidth=2 expandtab softtabstop=2 filetype=bash fileencoding=utf-8

[[ $TRACE ]] && set -x

# Check first if the required FTP_BUCKET variable was provided, if not, abort.
if [ -z $FTP_BUCKET ]; then
  echo "You need to set BUCKET environment variable. Aborting!"
  exit 1
fi

# Then check if there is an IAM_ROLE provided, if not, check if the AWS credentials were provided.
if [ -z $IAM_ROLE ]; then
  echo "You did not set an IAM_ROLE environment variable. Checking if AWS access keys where provided ..."
fi

# Abort if the AWS_ACCESS_KEY_ID was not provided if an IAM_ROLE was not provided neither.
if [ -z $IAM_ROLE ] &&  [ -z $AWS_ACCESS_KEY_ID ]; then
  echo "You need to set AWS_ACCESS_KEY_ID environment variable. Aborting!"
  exit 1
fi

# Abort if the AWS_SECRET_ACCESS_KEY was not provided if an IAM_ROLE was not provided neither. 
if [ -z $IAM_ROLE ] && [ -z $AWS_SECRET_ACCESS_KEY ]; then
  echo "You need to set AWS_SECRET_ACCESS_KEY environment variable. Aborting!"
  exit 1
fi

# If there is no IAM_ROLE but the AWS credentials were provided, then set them as the s3fs credentials.
if [ -z $IAM_ROLE ] && [ ! -z $AWS_ACCESS_KEY_ID ] && [ ! -z $AWS_SECRET_ACCESS_KEY ]; then
  #set the aws access credentials from environment variables
  echo $AWS_ACCESS_KEY_ID:$AWS_SECRET_ACCESS_KEY > ~/.passwd-s3fs
  chmod 600 ~/.passwd-s3fs
fi

# Update the vsftpd.conf file to include the IP address if running on an EC2 instance
if curl -s http://ifconfig.co > /dev/null ; then
  IP=$(curl -s http://ifconfig.co)
  sed -i "s/^pasv_address=.*/pasv_address=$IP/" /etc/vsftpd.conf
else
  exit 1
fi

if [[ ! -z $BANNER ]]; then
  sed -i "s/^ftpd_banner=.*/ftpd_banner=$BANNER/" /etc/vsftpd.conf
fi

# start s3 fuse
# Code above is not needed if the IAM role is attaced to EC2 instance 
# s3fs provides the iam_role option to grab those credentials automatically
MP_UMASK=0022
UMASK=0002
FS_OPTIONS="-o allow_other -o mp_umask=$MP_UMASK -o umask=$UMASK"

if [ ! -z $IAM_ROLE ]; then
  FS_OPTIONS="$FS_OPTIONS -o iam_role='$IAM_ROLE'"
fi

if [[ $TRACE ]] ; then
  FS_OPTIONS="$FS_OPTIONS -d -d -f -o f2 -o curldbg"
fi

/usr/local/bin/s3fs $FTP_BUCKET /home/aws/s3bucket $FS_OPTIONS

if [[ $? -gt 0 ]]; then
  echo "ERROR mounting '$FTP_BUCKET', cannot continue"
  exit 1
fi  

MOUNT_TIMEOUT=10
for ((i = 1; i <= $MOUNT_TIMEOUT; i++)); do
  if $(mount | egrep -q '^s3fs'); then
     echo "SUCCESS, mounted '$FTP_BUCKET'!"
     break
  else
    echo "No s3fs filesystem found, s3fs mount of '$FTP_BUCKET' is still connecting, waiting..."
    sleep 1
  fi
done

if ! $(mount | egrep -q '^s3fs'); then
  echo "No s3fs filesystem found, s3fs mount of '$FTP_BUCKET' probably failed"
  exit 2
fi

/usr/local/users.sh

